File: data-structures-algorithms-coursera/course1-algorithmic-toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/gcd.py
# Input: The two integers a, b are given in the same line separated by space.
# Output: The greatest common divisor of a and b

def gcd_naive(a, b):
    current_gcd = 1
    for d in range(2, min(a, b) + 1):
        if a % d == 0 and b % d == 0:
            if d > current_gcd:
                current_gcd = d

    return current_gcd


def gcd_euclid(a, b):
    if b == 0:
        return a
    a_prime = a % b
    return gcd_euclid(b, a_prime)


if __name__ == "__main__":
    a, b = map(int, input().split())
    # print(gcd_naive(a, b))
    print(gcd_euclid(a, b))

File: data-structures-algorithms-coursera/course1-algorithmic-toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/gcd.rs
// Input: The two integers a, b are given in the same line separated by space.
// Output: The greatest common divisor of a and b

use std::io;
use std::cmp;


fn gcd_naive(a: &u64, b: &u64) -> u64 {
    let mut current_gcd = 1;
    for i in 2..(cmp::min(a, b) + 1) {
        if a % i == 0 && b % i == 0{
            if i > current_gcd {
                current_gcd = i;
            }
        } 
    }
    current_gcd
}


fn gcd_euclid(a: &u64, b: &u64) -> u64 {
    if *b == 0 {
        return *a;
    }
    let a_prime = a % b;

    gcd_euclid(&b, &a_prime)
}


fn main() -> io::Result<()> {
    // get user inputs
    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer)?;
    let mut words = buffer.split_whitespace();
    let a: u64 = words.next().unwrap().parse().unwrap();
    let b: u64 = words.next().unwrap().parse().unwrap();

    // println!("{:?}", gcd_naive(&a, &b));
    println!("{:?}", gcd_euclid(&a, &b));

    Ok(())
}